brian2genn/templates/engine.cpp

```diff
@@ -108,46 +108,103 @@ void engine::run(double duration, //!< Duration of time to run the model for
+          {% else %}
+          convert_sparse_synapses_2_dynamic_arrays(C{{obj['owner'].name}}, {{var}}{{obj['owner'].name}},
+                                                   {{obj['srcN']}}, {{obj['trgN']}},
+                                                   brian::_dynamic_array_{{obj['owner'].name}}__synaptic_pre,
+                                                   brian::_dynamic_array_{{obj['owner'].name}}__synaptic_post,
+                                                   brian::_dynamic_array_{{obj['owner'].name}}_{{var}}, b2g::FULL_MONTY);
+          {% endif %}
+          {% else %}
+           copy_genn_to_brian({{var}}{{obj['owner'].name}},
+                              brian::_array_{{obj['owner'].name}}_{{var}},
+                              {{obj['owner'].variables[var].size}});
+          {% endif %}
+          {% endif %}
+          {% endfor %}
+
+          _run_{{obj['codeobj'].name}}();
+
+           {% for var in obj['write'] %}
+           {% if obj['isSynaptic'] %}
+           {% if obj['connectivity'] == 'DENSE' %}
+           convert_dynamic_arrays_2_dense_matrix(brian::_dynamic_array_{{obj['owner'].name}}__synaptic_pre,
+                                                 brian::_dynamic_array_{{obj['owner'].name}}__synaptic_post,
+                                                 brian::_dynamic_array_{{obj['owner'].name}}_{{var}},
+                                                 {{var}}{{obj['owner'].name}},
+                                                 {{obj['srcN']}}, {{obj['trgN']}});
+           {% else %}
+           convert_dynamic_arrays_2_sparse_synapses(brian::_dynamic_array_{{obj['owner'].name}}__synaptic_pre,
+                                                    brian::_dynamic_array_{{obj['owner'].name}}__synaptic_post,
+                                                    brian::_dynamic_array_{{obj['owner'].name}}_{{var}},
+                                                    {{var}}{{obj['owner'].name}},
+                                                    {{obj['srcN']}}, {{obj['trgN']}},
+                                                    _{{obj['owner'].name}}_bypre);
+           {% endif %}
+           {% else %}
+           copy_brian_to_genn(brian::_array_{{obj['owner'].name}}_{{var}},
+                              {{var}}{{obj['owner'].name}},
+                              {{obj['owner'].variables[var].size}});
+           {% endif %}
+           {% endfor %}
       }
       {% endif %}
       {% endfor %}
 #ifndef CPU_ONLY
       if (which == GPU) {
+          {% set states_pushed = [] %}
+          {% for run_reg in run_regularly_operations %}
+          {% for var in run_reg['write'] %}
+          {% if not run_reg['owner'].variables[var].owner.name in states_pushed %}
+          push{{run_reg['owner'].variables[var].owner.name}}StateToDevice();
+          {% if states_pushed.append(run_reg['owner'].variables[var].owner.name) %}{% endif %}
```

Comment:
Doesn't this mean that you are pushing target variables of run_regularly operations every timestep, even if they have not changed because i % {{obj['step']}} == 0 was false?

Reply:
Hah, indeed! I guess I was so focussed on not pushing the same state twice per time step that I forgot the obvious :)